sys/riscv/pmap.c

```diff
@@ -1,168 +1,708 @@
+/*
+ * Physical-to-virtual entries are managed for all pageable mappings.
+ */
+
+static void pv_add(pmap_t *pmap, vaddr_t va, vm_page_t *pg) {
+  assert(mtx_owned(&pv_list_lock));
+  pv_entry_t *pv = pool_alloc(P_PV, M_ZERO);
+  pv->pmap = pmap;
+  pv->va = va;
+  TAILQ_INSERT_TAIL(&pg->pv_list, pv, page_link);
+  TAILQ_INSERT_TAIL(&pmap->pv_list, pv, pmap_link);
+}
 
-  klog("Physical memory range: %p - %p", dmap_paddr_base, dmap_paddr_end - 1);
+static pv_entry_t *pv_find(pmap_t *pmap, vaddr_t va, vm_page_t *pg) {
+  assert(mtx_owned(&pv_list_lock));
+  pv_entry_t *pv;
+  TAILQ_FOREACH (pv, &pg->pv_list, page_link) {
+    if (pv->pmap == pmap && pv->va == va)
+      return pv;
+  }
+  return NULL;
+}
 
-  klog("dmap range: %p - %p", DMAP_VADDR_BASE, DMAP_VADDR_BASE + dmap_size - 1);
+static void pv_remove(pmap_t *pmap, vaddr_t va, vm_page_t *pg) {
+  assert(mtx_owned(&pv_list_lock));
+  pv_entry_t *pv = pv_find(pmap, va, pg);
+  TAILQ_REMOVE(&pg->pv_list, pv, page_link);
+  TAILQ_REMOVE(&pmap->pv_list, pv, pmap_link);
+  pool_free(P_PV, pv);
+}
 
-  /* Build direct map using superpages. */
-  pd_entry_t *pde = (void *)pd_va;
-  size_t idx = L0_INDEX(DMAP_VADDR_BASE);
-  for (paddr_t pa = dmap_paddr_base; pa < dmap_paddr_end; pa += L0_SIZE, idx++)
-    pde[idx] = PA_TO_PTE(pa) | PTE_KERN;
+/*
+ * Routines for accessing page table entries.
+ */
 
-  __sfence_vma();
+static paddr_t pmap_alloc_pde(pmap_t *pmap, vaddr_t va) {
+  assert(mtx_owned(&pmap->mtx));
 
-  void *fdtp = (void *)phys_to_dmap(FDT_get_physaddr());
-  FDT_changeroot(fdtp);
+  vm_page_t *pg = pmap_pagealloc();
+  assert(pg);
+
+  TAILQ_INSERT_TAIL(&pmap->pte_pages, pg, pageq);
+
+  klog("Page table for %p allocated at %p", va, pg->paddr);
+
+  return pg->paddr;
 }
 
-void pmap_enter(pmap_t *pmap, vaddr_t va, vm_page_t *pg, vm_prot_t prot,
-                unsigned flags) {
-  panic("Not implemented!");
+static pt_entry_t *pmap_lookup_pte(pmap_t *pmap, vaddr_t va) {
+  pd_entry_t *pdep = (pd_entry_t *)phys_to_dmap(pmap->pde);
+  pt_entry_t pte;
+
+  /* Level 0 */
+  pdep += L0_INDEX(va);
+  pte = *pdep;
+  if (!VALID_PDE_P(pte))
+    return NULL;
+
+  /* A direct map superpage? */
+  if (LEAF_PTE_P(pte))
+    return (pt_entry_t *)pdep;
+
+  pdep = (pd_entry_t *)phys_to_dmap(PTE_TO_PA(pte));
+
+  /* Level 1 */
+  return (pt_entry_t *)pdep + L1_INDEX(va);
 }
 
-bool pmap_extract(pmap_t *pmap, vaddr_t va, paddr_t *pap) {
-  panic("Not implemented!");
+static inline pd_entry_t make_pde(paddr_t pa) {
+  return PA_TO_PTE(pa) | PTE_V;
 }
 
-void pmap_remove(pmap_t *pmap, vaddr_t start, vaddr_t end) {
-  panic("Not implemented!");
+static pt_entry_t make_pte(paddr_t pa, vm_prot_t prot, unsigned flags,
+                           bool kernel) {
+  pt_entry_t pte = PA_TO_PTE(pa) | vm_prot_map[prot];
+
+  const pt_entry_t mask_on = (kernel) ? PTE_G : PTE_U;
+  pte |= mask_on;
+
+  const pt_entry_t mask_off = (kernel) ? 0 : PTE_D | PTE_A | PTE_W | PTE_V;
+  pte &= ~mask_off;
```

Comment:
You can remove `bool kernel` args and move that logic to `pmap_write_pte` like in `AArch64`.